communitynodes/FaceMask

ymgenesis opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update


## Have you discussed this change with the InvokeAI team?
- [x] Yes, briefly on discord. @psychedelicious @blessedcoolant 
- [ ] No, because:

      
## Description

The FaceMask node autodetects a face (and potentially other things) with an OpenCV cascade classifier and masks it by making a transparency in the image. You can swap faces with other faces, or swap things around the face with other things.

You provide it with an input image, and the full path to a OpenCV cascade classifier. I optimized the shape of the oval mask to work best with [haarcascade_frontalface_default.xml](https://github.com/kipr/opencv/blob/master/data/haarcascades/haarcascade_frontalface_default.xml), but many can be used. There are cascade classifiers already present at `invokeai/.venv/lib/python3.10/site-packages/cv2/data/`, as well. More information on OpenCV cascade classifiers: https://docs.opencv.org/3.4/db/d28/tutorial_cascade_classifier.html.

In addition to the image output with the newly-created transparency and HxW, the node also outputs an all-white mask in the same dimensions as the input image. This is needed by the inpaint node (and unified canvas) for outpainting; invokeai seems to pass in an all-white mask when it detects transparent pixels and initiates outpainting.

By default the node creates a vertically oriented oval shape to fit a classic face. The X and Y offset settings provide granular resizing/shaping of the oval mask for greater control of the masked area. If using another cascade classifier to detect a pair of eyes, for example, the offsets can be used to turn the ovals into circles.

## QA Instructions, Screenshots, Recordings

Trimmed out time while generations ran to shorten recording.

[Untitled2.webm](https://github.com/invoke-ai/InvokeAI/assets/25252829/db24e793-1820-49bd-9838-b81d11e31166)

ymgenesis commented:
Thinking I'm going to revert back to letting the user provide a full path to the cascade file... there's several new cascade files out there with different applications (anime, cat, etc) that aren't provided with invoke's cv2 implementation in the venv.

Or a new `cascades` folder can be added somewhere and users can drop the files in and they'll populate in the dropdown list. Just have to figure out how to populate it dynamically without explicitly creating the list of cascade files before the FaceMaskInvocation def.

Millu commented on invokeai/app/invocations/image.py:
```diff
@@ -59,6 +60,19 @@ class Config:
             ]
         }
 
+class ImageMaskOutput(BaseInvocationOutput):
```
Could you add the this to the list of base nodes in the docs?

https://github.com/invoke-ai/InvokeAI/blob/main/docs/features/NODES.md

ymgenesis replied:
Sure, I will make a PR to add FaceMask to the nodes docs.

ymgenesis commented:
Made some pretty drastic changes. Switched from OpenCV Cascade Classifier detection to MediaPipe detection. Removed cascade file input as a result. Also made a [pull request](https://github.com/invoke-ai/InvokeAI/pull/3876) for entry into communityNodes.md. 

I suppose this PR can be closed since community nodes will exist in a reference sheet.

hipsterusername commented:
@ymgenesis - With the latest communitynodes reference, I believe this can be closed, aye?

ymgenesis commented:
Sounds good.
